Lint uses of `ok().expect()`

fhartwig opened the pull request:
Fixes #429 (mostly).
This adds a lint for calling `ok().expect()` on a `Result`. Since I wanted to avoid false positives (if the error type doesn't implement `Debug`, you cannot call `expect` directly on the `Result`) and I couldn't figure out how to properly check if a type implements `Debug`, this has false negatives if the result's error type has type parameters. I don't think this is terribly common, though. If someone could point me in the right direction regarding checking for `Debug` impls, I'd be happy to fix this.

fhartwig commented:
Ok, this needs an update for the latest nighty. I'll look into that later today, unless someone else does it first.

Manishearth commented on src/methods.rs:
```diff
@@ -1,18 +1,81 @@
+#[derive(Clone)]
+pub struct MethodsPass { types_implementing_debug: Option<HashSet<DefId>> }
+
+impl MethodsPass {
```
Why record all the impls? Instead, just use http://manishearth.github.io/rust-internals-docs/rustc/middle/ty/struct.TraitDef.html#method.for_each_relevant_impl . If the iterator has nonzero length, you're done.

The traitdef already has a hashmap of impls in it, we're really just duplicating that.

fhartwig replied:
That does seem much more sensible, thanks :smiley:

fhartwig commented:
Ok, I've rebased this on master and replaced my trait-hackery with `for_each_relevant_impl`. We still have to check that the `relevant` impls are actually relevant (`for_each_relevant_impl` seems to always consider e.g. the `&T` impl relevant for some reason) and I still don't know how to make this work for generic error types, but this is definitely much nicer.

Manishearth commented on src/methods.rs:
```diff
@@ -115,6 +137,41 @@ impl LateLintPass for MethodsPass {
+    debug_def.for_each_relevant_impl(cx.tcx, no_ref_ty, |d| {
+        let self_ty = &cx.tcx.impl_trait_ref(d).and_then(|im| im.substs.self_ty());
+        if let Some(self_ty) = *self_ty {
+            if !self_ty.flags.get().contains(ty::TypeFlags::HAS_PARAMS) {
```
Do we need this check anymore? `Ty` contains parameter substs on both sides.

fhartwig replied:
I think we need some check to avoid matching blanket impls. Without this, any type will match the `Debug` impls for `&T` and `&mut T`, and if there is an impl for `Foo<T: Debug>`, `Foo<Whatever>` will match that impl, even if `Whatever` doesn't actually implement `Debug`. There may be a better way to do this, but I couldn't find one.

Manishearth commented:
Thanks!
